DM-6368: Allow Python 3 and do not assume SCons python is relevant

timj opened the pull request.

timj committed:
Allow python 3 to pass test

Uses bc command for floating point comparison.

Checks that minimum python3 version is 3.4 and minimum python2
version is 2.7.

timj committed:
Ignore _build scripts and .cfgc files

timj committed:
Use python from path not SCons python

We no longer know that the Python running SCons is the same as the
Python that we want to use when building and running the code.
This means the cfg file can no longer use sysconfig directly
but must ask the PATH python for that information.

jhoblitt commented on scripts/check_python:
```diff
@@ -11,8 +11,10 @@ if hash "$PYTHON" 2>/dev/null; then
-		echo "check-python: Python v2.7 not found (version reported by '$PYTHON' is $PYTHON_VERSION)." 2>&1
+	MINVER=2.7
+	MINVER3=3.4
+	if [ $(bc <<< "$PYTHON_VERSION < $MINVER || ($PYTHON_VERSION > 3.0 && $PYTHON_VERSION < $MINVER3)") -eq 1 ]; then
```
I had to really think about this one.  Is this copied from stack exchange or some place else that can be linked?

timj replied:
@PaulPrice had the same comment. `bc` seems to be the sanest way to do floating point comparisons. I did find an example somewhere when searching for "bash floating point comparisons" but I have no idea where that page is now. I'll add a comment stating that it's for floats.

jmatt replied:
It's just a really old python interface that can't change since it's related to python versions. See: https://docs.python.org/2/library/sys.html#sys.version_info and https://docs.python.org/3/library/sys.html#sys.version_info

jmatt replied:
The weird part to me is the script is bash instead of python.

timj replied:
I thought @jhoblitt was talking about the `bc` usage not the `sys.version_info` usage.

timj replied:
There's a lot of historical evolution of this script. The `numpy` and `matplotlib` version check scripts are Python. This one never has been, presumably because it's meant to be checking that python exists at all and then checking it's the right version. If you write it in python you have to also handle the difference in behavior when python is not there.

jmatt replied:
Yeah that is resolved if you just make it a python script and use the tuple. But `bc` works... and there are a lot of C developers on the team. I'd never use it but if this is acceptable to the rest of DM then I can follow it.

jhoblitt commented on ups/python.cfg:
```diff
@@ -11,19 +12,30 @@ class Configuration(lsst.sconsUtils.Configuration):
+        that we are using to run SCons so we must shell out to the
+        PATH python."""
+        pycmd = 'import distutils.sysconfig as s; print(s.get_config_var("{}"))'.format(name)
+        result = subprocess.check_output(["python", "-c", pycmd]).decode().strip()
```
If this raises an exception, will eups catch it or will everything exit?

timj replied:
You mean an error from decoding the bytes? SCons will be the thing that dies. Even if we caught it and returned `None` the compilation would break shortly afterwards.

jmatt commented on scripts/check_python:
```diff
@@ -11,8 +11,10 @@ if hash "$PYTHON" 2>/dev/null; then
-	if [[ "$PYTHON_VERSION" != "2.7" ]]; then
-		echo "check-python: Python v2.7 not found (version reported by '$PYTHON' is $PYTHON_VERSION)." 2>&1
+	MINVER=2.7
+	MINVER3=3.4
```
Is there any reason we're supporting 3.4? Instead of just 3.5? Is there value in limiting the versions of Python 3? Or on the opposite side, is there value is supporting more Python 3 versions?

timj replied:
Well, I picked a value. You are right that picking a high value now (that matches default anaconda) will reduce pain later on. I'll change to 3.5 and see what happens.

jmatt commented:
One high level comment that is addressed in this PR but I want to mention none the less. It's important to know when the code is breaking here. Especially for the unfortunate DM developer that isn't in the know and has a weird configuration. Specifically I have concerns with the Scons code which is likely to be reused in many places with arbitrary Scons configurations. All of that being said it works and there is no easy workaround. So it looks good.

timj committed:
Rewrite version test so it is done by python itself

Using shell scripts was becoming complicated given that testing
for 3.x and 2.x independently and having to support 3.10 being
newer than 3.9. Rewriting the logic in Python itself makes the
testing more obvious.

timj commented:
I've completely rewritten the version test. Please take another look.

@jmatt are you talking about the python.cfg file crashing with bad python data?

jmatt commented:
Excellent! I have no problem following that.
